src/sentry/attachments/base.py

```diff
@@ -25,60 +48,112 @@ def from_upload(cls, file, **kwargs):
+                timeout=timeout,
+                metrics_tags=metrics_tags,
+            )
+
+        meta = []
+
+        for attachment in attachments:
+            attachment._cache = self
+            meta.append(attachment.meta())
+
+        self.inner.set(ATTACHMENT_META_KEY.format(key=key), meta, timeout, raw=False)
+
+    def set_chunk(self, key, id, chunk_index, chunk_data, timeout=None):
+        key = ATTACHMENT_DATA_CHUNK_KEY.format(key=key, id=id, chunk_index=chunk_index)
+        self.inner.set(key, zlib.compress(chunk_data), timeout, raw=True)
+
+    def set_unchunked_data(self, key, id, data, timeout=None, metrics_tags=None):
+        key = ATTACHMENT_UNCHUNKED_DATA_KEY.format(key=key, id=id)
+        compressed = zlib.compress(data)
+        metrics.timing("attachments.blob-size.raw", len(data), tags=metrics_tags)
+        metrics.timing("attachments.blob-size.compressed", len(compressed), tags=metrics_tags)
+        metrics.incr("attachments.received", tags=metrics_tags, skip_internal=False)
+        self.inner.set(key, compressed, timeout, raw=True)
 
     def get(self, key):
-        key = self.make_key(key)
-        result = self.inner.get(key, raw=False)
+        result = self.inner.get(ATTACHMENT_META_KEY.format(key=key), raw=False)
         if result is not None:
-            result = [
-                CachedAttachment(
-                    load=lambda index=index: zlib.decompress(
-                        self.inner.get(u"{}:{}".format(key, index), raw=True)
-                    ),
-                    **attachment
-                )
-                for index, attachment in enumerate(result)
-            ]
+            rv = []
+            for id, attachment in enumerate(result):
+                attachment.setdefault("id", id)
+                rv.append(CachedAttachment(cache=self, key=key, **attachment))
+
+            return rv
+
         return result
 
+    def get_chunk(self, key, id, chunk_index):
+        key = ATTACHMENT_DATA_CHUNK_KEY.format(key=key, id=id, chunk_index=chunk_index)
+        return zlib.decompress(self.inner.get(key, raw=True))
+
+    def get_unchunked_data(self, key, id):
+        key = ATTACHMENT_UNCHUNKED_DATA_KEY.format(key=key, id=id)
+        result = self.inner.get(key, raw=True)
+        if result is None:
+            return result
+        return zlib.decompress(result)
+
     def delete(self, key):
-        key = self.make_key(key)
-        attachments = self.inner.get(key, raw=False)
+        attachments = self.inner.get(ATTACHMENT_META_KEY.format(key=key), raw=False)
         if attachments is None:
             return
 
-        for index in range(0, len(attachments)):
-            self.inner.delete(u"{}:{}".format(key, index))
+        for id, attachment in enumerate(attachments):
+            chunks = attachment.get("chunks")
+            if chunks is None:
+                self.inner.delete(ATTACHMENT_UNCHUNKED_DATA_KEY.format(key=key, id=id))
```

Comment:
You need to default to `attachment.id` here, and only use the index as fallback, to stay consistent with `set`. Can we refactor this into functions that receive the attachment model and return a list of all keys, to abstract this for `get`, `set`, `delete`, and so on?

Reply:
Attachment models now have a `cache_keys` property